### PR TITLE
Structs of Views instead of SoAs

### PR DESCRIPTION
**Pull request by PointKernel**
Hi Matti,
Here's the new data structure exploration. It's not completed yet cause I still get some segmentation faults during runtime. Otherwise, below listed all efforts I did and some issues to discuss:

- The original `Input` is an SoA containing some c-type arrays and another SoA (`SiPixelFedCablingMapGPU`). What I did is implementing an `InputKokkos<MemorySpace>` and it has two member variables `InputKokkosDevice<MemorySpace>` and `InputKokkosHost<MemorySpace>`. Device data are structs of views and host data is the corresponding host mirror.

- `KokkosInput` has member functions like `ReadInput()`, `DeepCopyD2H()` and `DeepCopyH2D()`. Compared to the original implementation where only one deep copy is required between the host mirror and the device view, here we need to copy each member individually.

- In this implementation, we have to template all related functions like `analyze()`, `rawtodigi()`, and all related functions and for now, I moved all functions to header files. One issue I cannot figure out right now is how to properly use `KOKKOS_NAMESPACE` instead of copy-paste three times the same function bodies for `kokkos_serial`, `kokkos_openmp` and `kokkos_cuda`. Maybe you know how to deal with this issue? 

- Since struct of built-in data types is supported by Kokkos view, I found the original implementation might be already a very good solution. For the code you showed before: https://github.com/cms-patatrack/pixeltrack-standalone/blob/master/src/kokkos/KokkosDataFormats/SiPixelDigisKokkos.h, I wonder how did you handle the host mirror in this case? Do you create another corresponding host mirror class as well?

- Voilà, I will take a look at the segfault tomorrow. Hopefully it's a small issue.

### CODE REVIEW

**Comment by PointKernel** [comment]
Please do not merge for now. Thanks.

File: kokkosConfig.h [hunk 1]
```diff
@@ -5,12 +5,15 @@
 
 #ifdef DIGI_KOKKOS_SERIAL
 using KokkosExecSpace = Kokkos::Serial;
+using KokkosMemSpace = Kokkos::HostSpace;
```
Comment (makortel) [inline kokkosConfig.h]: You can actually pass the `KokkosExecSpace` as the memory space template argument, so I'd expect these type aliases to be unnecessary (at least to leading order).

**Comment by makortel** [comment]
Thanks Yunsong. I have some comments below.

> * What I did is implementing an `InputKokkos<MemorySpace>` and it has two member variables `InputKokkosDevice<MemorySpace>` and `InputKokkosHost<MemorySpace>`

I was (silently) hoping for a solution that would have had only one `InputKokkos<...>` definition, but I don't have good ideas now how to do that actually (maybe with some template magic).

> * One issue I cannot figure out right now is how to properly use `KOKKOS_NAMESPACE` instead of copy-paste three times the same function bodies for `kokkos_serial`, `kokkos_openmp` and `kokkos_cuda`. Maybe you know how to deal with this issue?

Could you elaborate what kind of problems you encountered? The `KOKKOS_NAMESPACE` is defined in https://github.com/makortel/pixel-standalone/blob/master/kokkosConfig.h according to which backend macro is defined in the compiler flags.

> * Since struct of built-in data types is supported by Kokkos view, I found the original implementation might be already a very good solution. For the code you showed before: https://github.com/cms-patatrack/pixeltrack-standalone/blob/master/src/kokkos/KokkosDataFormats/SiPixelDigisKokkos.h, I wonder how did you handle the host mirror in this case? Do you create another corresponding host mirror class as well?

I have not handled it yet in any way :) The `SiPixelDigisKokkos` is filled on the device, and in the scope of that project does not have to be transferred to host.

For the calibration data I filled and `deep_copy()`d each `HostMirror` column explicitly after which they get discarded, see e.g.
https://github.com/cms-patatrack/pixeltrack-standalone/blob/master/src/kokkos/plugin-SiPixelClusterizer/kokkos/SiPixelGainCalibrationForHLTESProducer.cc
https://github.com/cms-patatrack/pixeltrack-standalone/blob/master/src/kokkos/CondFormats/SiPixelGainForHLTonGPU.h

**Comment by makortel** [comment]
The changes are rather invasive, and it is indeed unclear if this is a better solution than the existing `View` of a struct, so I'd prefer to create a new set of files instead of modifying the existing `kokkos` files, e.g. `kokkosview` (with `DIGI_KOKKOSVIEW` macro and all the copy-paste in `Makefile`). But that can be done as a last step after figuring out all the troubles.

File: kokkosConfig.h (hunk 1 above)
Comment (PointKernel) [reply kokkosConfig.h]: sure, will do.

Comment (PointKernel) [reply kokkosConfig.h]: fixed

**Comment by PointKernel** [comment]
Now it compiles and runs without any issue. What do you think?

File: input.h [hunk 2]
```diff
@@ -16,12 +21,124 @@ inline Input read_input() {
...
+  const SiPixelFedCablingMapKokkosDevice<MemorySpace>& getCablingMap() const { return this->cablingMap; }
+  KOKKOS_INLINE_FUNCTION
+  const Kokkos::View<unsigned char*, MemorySpace>& getFedId() const { return this->fedId; }
+  KOKKOS_INLINE_FUNCTION
+  const Kokkos::View<unsigned int*, MemorySpace>& getWord() const { return this->word; }
+  KOKKOS_INLINE_FUNCTION
+  const Kokkos::View<unsigned int*, MemorySpace>& getWordCounter() const { return this->wordCounter; }
+
+private:
+private:
```
Comment (makortel) [inline input.h]:
```suggestion
private:
```

File: pixelgpudetails.h [hunk 3]
```diff
@@ -3,6 +3,10 @@
 
 #include <cstdint>
 
+#ifdef DIGI_KOKKOS
```
Comment (makortel) [inline pixelgpudetails.h]:
```suggestion
#ifdef DIGI_KOKKOSVIEW
```

File: rawtodigi_kokkosview.cc [hunk 4]
```diff
@@ -0,0 +1 @@
+#include "rawtodigi_kokkos.h"
```
Comment (makortel) [inline rawtodigi_kokkosview.cc]: If everything is in the header, why is this file needed?

File: rawtodigi_kokkosview.h [hunk 5]
```diff
@@ -0,0 +1,494 @@
+#ifndef rawtodigi_kokkos_h_
+#define rawtodigi_kokkos_h_
+
+#include <Kokkos_Macros.hpp>
+
+#include "kokkosConfig.h"
+#include "input.h"
+#include "output.h"
+
+namespace kokkos_serial {
```
Comment (makortel) [inline rawtodigi_kokkosview.h:12]: I'd like to have a comment for a future reader (if any)
```suggestion
// In principle the namespace should be KOKKOS_NAMESPACE, but for some reason that lead to weird compilation
// errros. To be investigated later.
namespace kokkos_serial {
```

**Review by makortel — COMMENTED** [review]
Thanks, overall looks good now. I found a few minor things that could be improved.

File: input.h (hunk 2 above)
Comment (PointKernel) [reply input.h]: Done. Thanks for the thorough check.

File: pixelgpudetails.h (hunk 3 above)
Comment (PointKernel) [reply pixelgpudetails.h]: Done

File: rawtodigi_kokkosview.cc (hunk 4 above)
Comment (PointKernel) [reply rawtodigi_kokkosview.cc]: Yesh, at one point I didn't want to touch the Makefile so just took this easy solution. Fixed now.

File: rawtodigi_kokkosview.h (hunk 5 above)
Comment (PointKernel) [reply rawtodigi_kokkosview.h:12]: Added
